### PR TITLE
RDX: Split extension tag from ID

### LINKED ISSUE #4416: RDX: Remove image tag from extension ID
### Problem Description

Currently, we use the full image reference (including the tag) for the extension ID. This will not work well with upgrades, and makes it possible to install two versions of the same extension concurrently.

### Proposed Solution

We should split the extension ID to exclude the tag, and add a new field to hold the tag.

### Additional Information

_No response_

#### Acceptance Criteria

- Try to install an extension ID specifying a tag. RDX must install the requested tag.
- Try to install an extension ID with a tag that does not exist in the registry. RDX must throw an error.
- Try to install an extension ID that is already installed with a different tag. RDX will automatically uninstall the existing tag and install the requested tag. Different tags will be stored in the container engine, but only the latter will be the active version in the UI.

### PR DESCRIPTION
**Pull request by mook-as**
This splits out the extension tag from the extension ID, so that
- We only install one version of the extension at a time (instead of possibly installing multiple versions concurrently)
- This will let us later implement `rdctl extension upgrade`
- When we try to fetch an extension without a tag, we now get the currently-installed version rather than "latest".

Fixes #4416

### CODE REVIEW

File: background.ts [hunk 1]
```diff
@@ -504,8 +503,29 @@ ipcMainProxy.on('preferences-set-dirty', (_event, dirtyFlag) => {
 });
 
 function writeSettings(arg: RecursivePartial<RecursiveReadonly<settings.Settings>>) {
-  // arrayCustomizer is necessary to properly merge array of strings
-  _.mergeWith(cfg, arg, arrayCustomizer);
+  const customizer = (objValue: any, srcValue: any) => {
```
Comment (ericpromislow) [inline background.ts:506]: Thinking that if this whole block got moved into `utils/object.js` or a newer function we can unit-test it more easily. I get what it's doing, but think it would be worth having unit tests for it because there are so many edge cases.

File: pkg/rancher-desktop/main/extensions/extensions.ts [hunk 2]
```diff
@@ -324,11 +334,24 @@ export class ExtensionImpl implements Extension {
       {
         name:      this.containerName,
         namespace: this.extensionNamespace,
-        env:       { DESKTOP_PLUGIN_IMAGE: this.id },
+        env:       { DESKTOP_PLUGIN_IMAGE: this.image },
       },
     );
   }
 
+  get isInstalled(): Promise<boolean> {
```
Comment (ericpromislow) [inline pkg/rancher-desktop/main/extensions/extensions.ts:342]: I don't see the advantage of making this a getter. The name assumes it's a method, and making it async makes it more method-like. If getters' return values were cached I'd say otherwise, but they aren't, right?

File: pkg/rancher-desktop/main/commandServer/settingsValidator.ts [hunk 3]
```diff
@@ -492,6 +492,79 @@ export default class SettingsValidator {
...
+  ): boolean {
+    if (typeof desiredValue !== 'object' || !desiredValue) {
+      errors.push(`${ fqname }: "${ desiredValue }" is not a valid mapping`);
+
+      return false;
+    }
+
+    // makeRE is a tagged template for making regular expressions with /x (i.e.
+    // ignoring any whitespace within the regular expression itself).
+    function makeRE(strings: {raw: readonly string[]}, ...subsitutions: any[]) {
```
Comment (ericpromislow) [inline pkg/rancher-desktop/main/commandServer/settingsValidator.ts:510]: typo: should be `substitutions`

File: pkg/rancher-desktop/main/commandServer/settingsValidator.ts [hunk 4]
```diff
@@ -492,6 +492,79 @@ export default class SettingsValidator {
...
+      const subsitutionSources = subsitutions.map(s => s instanceof RegExp ? s.source : s);
+      const raw = String.raw(strings, ...subsitutionSources);
+      const lines = raw.split(/\r?\n/);
+      // Drop comments at end of line
+      const uncommentedLines = lines.map(line => line.replace(/\s#.*$/, ''));
```
Comment (ericpromislow) [inline pkg/rancher-desktop/main/commandServer/settingsValidator.ts:515]: So you'll never backslash-escape a `#` I take it

Comment (mook-as) [reply pkg/rancher-desktop/main/commandServer/settingsValidator.ts:515]: It only matches `#` prefixed by whitespace. So if you actually need it in a RE like that you can always write `[#]` to get around it trivially.

**Comment by ericpromislow** [comment]
During rdctl e2e tests I got a popup error message with this info:
```
Container rd-extension-rd_extension_everything-r-d-x-port-forwarding-1 Waiting
Container rd-extension-rd_extension_everything-backend-service-1 Healthy
container rd-extension-rd_extension_everything-r-d-x-port-forwarding-1 exited (2)
...
Last command run:
/Users/ericp/workspace/rancher/desktop/resources/darwin/bin/docker-compose --project-directory /Users/ericp/Library/Application Support/rancher-desktop/extensions/cmQvZXh0ZW5zaW9uL2V2ZXJ5dGhpbmc/compose --project-name rd-extension-rd_extension_everything up --quiet-pull --wait --remove-orphans
```

Didn't see anything useful in the logfile section of the popup error message

File: pkg/rancher-desktop/main/commandServer/settingsValidator.ts (hunk 3 above)
Comment (ericpromislow) [reply pkg/rancher-desktop/main/commandServer/settingsValidator.ts:510]: Also I have to ask, so it's probably worth commenting on what exactly `strings` is, because I'm not parsing the `{raw: readonly string[]}` type annotation

File: pkg/rancher-desktop/main/commandServer/settingsValidator.ts [hunk 5]
```diff
@@ -492,6 +492,79 @@ export default class SettingsValidator {
...
+      errors.push(`${ fqname }: "${ desiredValue }" is not a valid mapping`);
+
+      return false;
+    }
+
+    // makeRE is a tagged template for making regular expressions with /x (i.e.
+    // ignoring any whitespace within the regular expression itself).
+    function makeRE(strings: {raw: readonly string[]}, ...subsitutions: any[]) {
+      const subsitutionSources = subsitutions.map(s => s instanceof RegExp ? s.source : s);
+      const raw = String.raw(strings, ...subsitutionSources);
```
Comment (ericpromislow) [inline pkg/rancher-desktop/main/commandServer/settingsValidator.ts:512]: Regarding this line, my IDE says:
```
<html>TS2345: Argument of type '{ raw: readonly string[]; }' is not assignable to parameter of type 'TemplateStringsArray'.

Type '{ raw: readonly string[]; }' is missing the following properties from type 'TemplateStringsArray': length, concat, join, slice, and 18 more.
```

So my IDE is having as much trouble as I am understanding that def'n.

**Comment by ericpromislow** [comment]
e2e failure on extensions.e2e.spec.ts:
```
  1) extensions.e2e.spec.ts:338:7 › Extensions › extension API › ddClient.extension.vm.service › can fetch from the backend

    Error: expect(received).resolves.toContain()

    Received promise rejected instead of resolved
    Rejected to value: [Error: jsHandle.evaluate: Error: Error invoking remote method 'extensions/vm/http-fetch': FetchError: request to http://127.0.0.1:49153/etc/os-release failed, reason: connect ECONNREFUSED 127.0.0.1:49153]
```

File: pkg/rancher-desktop/main/commandServer/settingsValidator.ts [hunk 6]
```diff
@@ -492,6 +492,79 @@ export default class SettingsValidator {
     return false;
   }
 
+  protected checkExtensions(
```
Comment (ericpromislow) [inline pkg/rancher-desktop/main/commandServer/settingsValidator.ts:495]: Could you add some unit tests in `pkg/rancher-desktop/main/commandServer/__tests__/settingsValidator.spec.ts` on this?

**Review by ericpromislow — CHANGES_REQUESTED** [review]
One test failure, one typo, and a few questions and style/testing suggestions.

File: pkg/rancher-desktop/main/extensions/manager.ts [hunk 7]
```diff
@@ -206,58 +212,79 @@ class ExtensionManagerImpl implements ExtensionManager {
...
+    }
   }
 
-  getExtension(id: string): Extension {
-    let ext = this.extensions[id];
+  async getExtension(image: string): Promise<Extension> {
+    let [, repo, tag] = /^(.*):(.*?)$/.exec(image) ?? ['', image, undefined];
+    const extGroup = this.extensions[image] ?? {};
+
+    // The build process uses an older TypeScript that can't infer id correctly.
```
Comment (ericpromislow) [inline pkg/rancher-desktop/main/extensions/manager.ts:239]: I assume `id` is referring to obsolete code

File: pkg/rancher-desktop/main/commandServer/settingsValidator.ts [hunk 8]
```diff
@@ -492,6 +492,83 @@ export default class SettingsValidator {
...
+
+      return false;
+    }
+
+    /**
+     * makeRE is a tagged template for making regular expressions with /x (i.e.
+     * ignoring any whitespace within the regular expression itself).
+     * @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals#tagged_templates
+     */
+    function makeRE(strings: TemplateStringsArray, ...substitutions: any[]) {
```
Comment (ericpromislow) [inline pkg/rancher-desktop/main/commandServer/settingsValidator.ts:518]: The `makeRE` function and all uses of it are constants and can be hoisted outside this instance method

**Review by ericpromislow — CHANGES_REQUESTED** [review]
Two small changes:
- hoist `makeRE` etc. out of the instance method
- Fix a comment referring to `id` instead of `images`

File: pkg/rancher-desktop/main/commandServer/settingsValidator.ts (hunk 8 above)
Comment (mook-as) [reply pkg/rancher-desktop/main/commandServer/settingsValidator.ts]: Moved to global scope in 0d30157beb3c1da83b549608177f080269872f57.

File: pkg/rancher-desktop/main/commandServer/settingsValidator.ts (hunk 3 above)
Comment (ericpromislow) [reply pkg/rancher-desktop/main/commandServer/settingsValidator.ts]: Just a note that `raw` is not the best name to allude to everything this method does

**Review by ericpromislow — APPROVED** [review]
LGTM

File: pkg/rancher-desktop/main/commandServer/settingsValidator.ts (hunk 3 above)
Comment (mook-as) [reply pkg/rancher-desktop/main/commandServer/settingsValidator.ts]: Unfortunately, it's called `raw` because that's what the JS API passes in. (In the mean time, this has been updated to use the `TemplateStringsArray` interface directly, to better show that we don't define the shape of this argument.)
